### PR TITLE
[ Milestone 6 ] Preserve data

### PR DESCRIPTION
**Pull request by Harrynoqs**
- On exit from your app save all the data (books, people, and rentals) into JSON files. One per type of data you are storing and named accordingly (e.g. rentals.json)
- On startup of your app load all data from files.

### CODE REVIEW

File: app.rb [hunk 1]
```diff
@@ -12,11 +15,13 @@ def initialize
     @books = []
     @people = []
     @rentals = []
+    load_data
```
Comment (ShadyShawkat) [inline app.rb:17]: - [x] Please make sure that the app can function properly with no errors shown if there are one or more missing files, as per [this](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/school_library_preserve_data.md#:~:text=Remember%20to%20handle%20the%20case%20when%20one%20or%20more%20of%20the%20files%20are%20missing) requirement.

image

File: app.rb [hunk 2]
```diff
@@ -25,126 +30,163 @@ def list_all_books
...
+      @people << person
+    end
+
+    rentals_data = JSON.parse(File.read('rentals.json'))
+    rentals_data.each do |rental|
+      date = rental['date']
+      person = @people.find { |p| p.id == rental['person_id'] }
+      book = @books.find { |b| b.id == rental['book_id'] }
+      @rentals << Rental.new(date, person, book)
+    end
```
Comment (ShadyShawkat) [inline app.rb]: - [x] Kindly make sure that the rentals are preserved correctly after the app is reopened. Currently, when I try to save a rental and show it after I reopen the app, I get the following result:

image

**Review by ShadyShawkat — CHANGES_REQUESTED** [review]
## Hi Team, 

Good job so far! 👍🏻 
There are some issues that you still need to work on to go to the next project but you are almost there!

## To Highlight
- All linters are passing ✔️ 
- Correct use of GitFlow ✔️ 
- Descriptive commit messages ✔️ 
- Professional looking pull request ✔️ 

Image

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app.rb (hunk 2 above)
Comment (CakeInTech) [reply app.rb]: - It's fixed now rentals is preserved correctly when reopened and you can get the saved person id by listing all people. 😄

File: app.rb (hunk 1 above)
Comment (CakeInTech) [reply app.rb:17]: - Created a condition for when the json file doesn't exist. 😄

**Review by christianbiring1 — APPROVED** [review]
Hi @MohamedCK  @Harrynoqs  , 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 
# STATUS: APPROVED :tada: 

## Highlights
- [x] The data can be saved correctly after the app is closed :+1: 
- [x] All linters checks are green :heavy_check_mark: 
## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**